hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java

```diff
@@ -768,7 +768,11 @@ public boolean next(List<Cell> outResult, ScannerContext scannerContext) throws
         // One last chance to break due to size limit. The INCLUDE* cases above already check
         // limit and continue. For the various filtered cases, we need to check because block
         // size limit may have been exceeded even if we don't add cells to result list.
-        if (scannerContext.checkSizeLimit(LimitScope.BETWEEN_CELLS)) {
+        // And when reaching the heartbeat cells, try to return from the loop.
+        if (
+          scannerContext.checkSizeLimit(LimitScope.BETWEEN_CELLS)
+            || kvsScanned % cellsPerHeartbeatCheck == 0
```

Comment:
should this only apply to `!matcher.isUserScan()`?

Reply:
We have discussed about the feature applying to compaction scan or user scan in https://issues.apache.org/jira/browse/HBASE-25709. Actually, we want the scans not stuck/loop in skip cells, and both the user and the compaction scans should return as soon as possible.

Reply:
After HBASE-27558 that you added size limit checker in the loop, it can applies to only compaction scans now.
`if (scannerContext.checkSizeLimit(LimitScope.BETWEEN_CELLS)) {
          return scannerContext.setScannerState(NextState.MORE_VALUES).hasMoreValues();
        }`

Reply:
I wonder actually if all of this can be simplified based on my change in HBASE-27558 -- What if we just did `scannerContext.setSizeLimit(ScannerContext.LimitScope.BETWEEN_CELLS, Long.MAX_VALUE, Long.MAX_VALUE, 100*1024*1024)` in Compactor?  Force the StoreScanner to return every 100mb or so of blocks scanned.

Reply:
I think the `kvsScanned % cellsPerHeartbeatCheck` was meant to avoid the cost of checking time limit (which requires hitting the clock). You're re-using it here as well to force compaction to check-in periodically based on cells/bytes scanned. But that's actually the whole point behind the changes in HBASE-27558, so that makes me think we should just apply the same change to compactions by specificying a block size limit in Compator.

Reply:
Seems good idea. It's reasonable to change the solution after the headway in [HBASE-27558](https://issues.apache.org/jira/browse/HBASE-27558). Thanks.